spec/connection.md

```diff
@@ -0,0 +1,229 @@
+# NATS Connection
+
+|Metadata|Value|
+|--------|-----|
+|Date    |2023-10-12 |
+|Author  |@Jarema |
+|Status  |`Implemented`|
+|Tags    |client, server|
+
+|Revision|Date|Author|Info|
+ |--------|----|------|----|
+ |1       |2023-10-12|@author|Initial draft|
+
+## Summary
+
+This document describes how clients connect to the NATS server or NATS cluster.
+That includes topics like:
+- connection process
+- reconnect
+- tls
+- discoverability of other nodes in a cluster
+
+## Motivation
+
+Ensuring a consistent way how Clients establish and maintain connection with the NATS server and provide consistent and predictable behaviour across the ecosystem.
+
+## Guide-level Explanation
+
+### Establishing the connection
+
+#### Minimal example
+
+1. Clients initiate a TCP connection to the Server.
+2. Server responds with [INFO][INFO] json.
+3. Client sends [CONNECT][CONNECT] json.
+4. Clients and Server start to exchange PING/PONG messages to detect if the connection is alive.
+
+#### Auth flow
+TODO
+
+#### TLS
+There are two flows available in the Server that enable TLS.
+
+##### Standard TLS
+
+This method is available in all NATS Server versions.
+
+1. Clients initiate a TCP connection to the Server.
+2. Server responds with [INFO][INFO] json.
+3. If Server [INFO][INFO] contains `tls_required` set to `true`, or the client has a tls requirement set to `true`, the client performs a TLS upgrade.
+4. Client sends [CONNECT][CONNECT] json.
+5. Clients and Server start to exchange PING/PONG messages to detect if the connection is alive.
+
+##### TLS First
+
+This method has been available since NATS Server 2.11.
+
+There are two prerequisites to use this method:
+1. Server config has enabled `handshake_first` field in the `tls` block.
+2. The client has set the `tls_first` option set to true.
+
+**handshake_first**
+has those possible values:
+- **`false`**: handshake first is disabled. Default value
+- `true`: handshake first is enabled and enforced. Clients that do not use this flow will fail to connect.
+- `duration` (i.e. 2s): a hybrid mode that will wait a given time, allowing the client to follow the `tls_first` flow. After the duration has expired, `INFO` is sent, enabling standard client TLS flow.
+- `auto`: same as above, with some default value. By default it waits 50ms for TLS upgrade before sending the [INFO][INFO].
```

Comment:
50ms seems super low to me

Reply:
that what is done in Server PR.
https://github.com/nats-io/nats-server/pull/4642

Reply:
Derek picked this value (was even going for 25ms). I too think that this is a bit low, but since it is configurable, it is less of an issue. And to be fair, I am expecting users to put their own duration. It is not really "saving" much in term of configuration for the user to write "auto" or an actual duration.